src/protocols/rdp/channels/cliprdr.c

```diff
@@ -0,0 +1,605 @@
+        CliprdrClientContext* cliprdr, UINT32 format) {
+
+    /* FreeRDP-specific handlers for CLIPRDR are not assigned, and thus not
+     * callable, until after the relevant guac_rdp_clipboard structure is
+     * allocated and associated with the CliprdrClientContext */
+    guac_rdp_clipboard* clipboard = (guac_rdp_clipboard*) cliprdr->custom;
+    assert(clipboard != NULL);
+
+    /* Create new data request */
+    CLIPRDR_FORMAT_DATA_REQUEST data_request = {
+        .requestedFormatId = format
+    };
+
+    /* Note the format we've requested for reference later when the requested
+     * data is received via a Format Data Response PDU */
+    clipboard->requested_format = format;
+
+    guac_client_log(clipboard->client, GUAC_LOG_TRACE, "CLIPRDR: Sending "
+            "format data request.");
+
+    /* Send request */
+    return cliprdr->ClientFormatDataRequest(cliprdr, &data_request);
+
+}
+
+/**
+ * Returns whether the given Format List PDU indicates support for the given
+ * clipboard format.
+ *
+ * @param format_list
+ *     The CLIPRDR_FORMAT_LIST structure representing the Format List PDU
+ *     being tested.
+ *
+ * @param format_id
+ *     The ID of the clipboard format to test, such as CF_TEXT or
+ *     CF_UNICODETEXT.
+ *
+ * @return
+ *     Non-zero if the given Format List PDU indicates support for the given
+ *     clipboard format, zero otherwise.
+ */
+static int guac_rdp_cliprdr_format_supported(const CLIPRDR_FORMAT_LIST* format_list,
+        UINT format_id) {
+
+    /* Search format list for matching ID */
+    for (int i = 0; i < format_list->numFormats; i++) {
+        if (format_list->formats[i].formatId == format_id)
+            return 1;
+    }
+
+    /* If no matching ID, format is not supported */
+    return 0;
+
+}
+
+/**
+ * Callback invoked by the FreeRDP CLIPRDR plugin for received Format List
+ * PDUs. The Format List PDU is sent by the RDP server to indicate that new
+ * clipboard data has been copied and is available for retrieval in the formats
+ * listed. A client wishing to retrieve that data responds with a Format Data
+ * Request PDU.
+ *
+ * @param cliprdr
+ *     The CliprdrClientContext structure used by FreeRDP to handle the CLIPRDR
+ *     channel for the current RDP session.
+ *
+ * @param format_list
+ *     The CLIPRDR_FORMAT_LIST structure representing the Format List PDU that
+ *     was received.
+ *
+ * @return
+ *     CHANNEL_RC_OK (zero) if the PDU was handled successfully, an error code
+ *     (non-zero) otherwise.
+ */
+static UINT guac_rdp_cliprdr_format_list(CliprdrClientContext* cliprdr,
+        const CLIPRDR_FORMAT_LIST* format_list) {
+
+    /* FreeRDP-specific handlers for CLIPRDR are not assigned, and thus not
+     * callable, until after the relevant guac_rdp_clipboard structure is
+     * allocated and associated with the CliprdrClientContext */
+    guac_rdp_clipboard* clipboard = (guac_rdp_clipboard*) cliprdr->custom;
+    assert(clipboard != NULL);
```

Comment:
This is more a curiosity than anything, but is there any special reason to using `assert` here instead of `if (clipboard == NULL); return`?  I'm familiar with `assert` in the Java side of the code, but it seems pretty rare in Guacamole's C code-based, so I'm just curious why it goes here and how it changes behavior?

Reply:
`assert()` in C will cause SIGABRT to be sent to the process if the condition fails. If we were expecting `clipboard` to occasionally be `NULL`, then adding an `if` would be best here. In this case, it is reasoned that `clipboard` absolutely cannot be `NULL` under any circumstance where this code can be reached, and we declare that assumption with an `assert()`. Smart compilers, static checkers, etc. will take that `assert()` into account, and it allows the system to bail out if a required assumption ends up being false.

Reply:
Thx.